### PR TITLE
[SPARK-41530][CORE] Rename MedianHeap to PercentileMap and support percentile

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

`MedianHeap` was added to track the median of task durations, for task speculation. However, median may not be the best option and a configurable percentile could be better.

This PR extends the existing `MedianHeap` to support percentile.

### Why are the changes needed?

Prepare for tracking more statistics of task durations.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @Ngone51

**Comment by mridulm** [comment]
> Prepare for tracking more statistics of task durations.

Can you give more details on this please ?

File: core/src/test/scala/org/apache/spark/util/collection/PercentileHeapSuite.scala [hunk 1]
```diff
@@ -21,46 +21,67 @@ import java.util.NoSuchElementException
 
 import org.apache.spark.SparkFunSuite
 
-class MedianHeapSuite extends SparkFunSuite {
+class PercentileHeapSuite extends SparkFunSuite {
 
-  test("If no numbers in MedianHeap, NoSuchElementException is thrown.") {
-    val medianHeap = new MedianHeap()
+  test("If no numbers in PercentileHeap, NoSuchElementException is thrown.") {
+    val medianHeap = new PercentileHeap()
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/util/collection/PercentileHeapSuite.scala:27]: Shall we rename the variable name too, `val medianHeap` -> `val percentileHeap`?

**Review by dongjoon-hyun — COMMENTED** [review]
Technically, `MedianHeap` is removed by being superceded by `PercentileHeap` after this PR.
- Since new class is important than the removed one, could you revise the PR title to mention `PercentileHeap` explicitly?
- Also, in the test suite, I believe it would be better not to mention `mediaHeap`.

**Comment by cloud-fan** [comment]
> Can you give more details on this please ?

Using the median task duration to trigger task speculation may not be the best option. It's better to allow users to configure the percentile if they want task speculation to happen more or less likely.

**Comment by mridulm** [comment]
> Using the median task duration to trigger task speculation may not be the best option. It's better to allow users to configure the percentile if they want task speculation to happen more or less likely.

Sounds good.
Note that given the complexity of how we determine whether a task is speculatable  or not, it is not just a matter of turning the dial on the percentile :-)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @cloud-fan .

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4. Thank you, @cloud-fan and @mridulm .

**Pull request closed without merging**